### PR TITLE
0089/stats object

### PR DESCRIPTION
**Pull request by davbyr**
After fighting with this for a while, I've settled on some decisions regarding different observations types and comparing them directly (MAE etc) to the model. Difficulties were to do with me trying to make a general STATS object to do this with any observations object. I've abandoned this and adopted a observation specific approach instead. 

There is no new object. Instead, there is a method inside altimetry called `obs_operator()`, which interpolates a given variable from a model object onto the same times and locations as the altimetry object. The interpolated array is of the same form as the altimetry variable arrays and so is stored inside the altimetry object's Dataset. Each observation type would have this method to deal with different data formats.

There's a few new methods in the NEMO class which are used for this: `nearest_xy_indices()`, `interpolate_in_space()` and `interpolate_in_time()`:

1. `nearest_xy_indices()` is similar to the existing `find_i_j()` but uses BallTree (faster) and can handle arrays of floats. 
2. `interpolate_in_space()` just finds the above indices and applies them for convenience.
3. `interpolate_in_time()` uses xarray.interp() to interpolate a dataset in time. It first renames `t_dim` to `time` so that it is a coordinate dimension.

These methods could eventually go in the MODEL class.

### CODE REVIEW

**Comment by jpolton** [comment]
@anwiseNOCL Need to install sklearn package:
`conda install -c intel scikit-learn`

File: coast/__init__.py [hunk 1]
```diff
@@ -8,6 +8,5 @@
 from .CDF import CDF
 from .CRPS import CRPS
 from .interpolate_along_dimension import interpolate_along_dimension
-from .STATS import STATS
+from .package import PACKAGE
```
Comment (jpolton) [inline coast/__init__.py:11]: Line 11: I don't think that you want PACKAGE? I commented it out to get it to work.

File: unit_testing/unit_test.py [hunk 2]
```diff
@@ -501,16 +501,16 @@
...
-    if (altimetry_nwes.dataset.dims['time'] == 213) :
+    if (altimetry_nwes.dataset.dims['t_dim'] == 213) :
         print(str(sec) + chr(subsec) + " OK - ALTIMETRY object subsetted using isel ")
     else:
         print(str(sec) + chr(subsec) + "X - Failed to subset object/ return as copy")
 except:
     print(str(sec) + chr(subsec) +" FAILED")
 #################################################
-## ( 6 ) CRPS Methods                         ##
+## ( 6 ) Validation Methods                    ##
```
Comment (jpolton) [inline unit_testing/unit_test.py:513]: You need to change the `unit_test_contents.txt` file too

File: unit_testing/unit_test.py [hunk 3]
```diff
@@ -562,6 +562,27 @@
     print(str(sec) + chr(subsec) + " OK - CRPS CDF plot saved")
 except:
     print(str(sec) + chr(subsec) + " X - CRPS CDF plot not saved")
+    
+#-----------------------------------------------------------------------------#
+# ( 6b ) Interpolate model to altimetry                                       #
```
Comment (jpolton) [inline unit_testing/unit_test.py:567]: 6c

File: unit_testing/unit_test.py [hunk 4]
```diff
@@ -562,6 +562,27 @@
     print(str(sec) + chr(subsec) + " OK - CRPS CDF plot saved")
```
Comment (jpolton) [inline unit_testing/unit_test.py:562]: I notice you have two items in 6b. This looks odd. Perhaps at line 562 and 564
```print(str(sec) + chr(subsec) + " OK - CRPS CDF plot saved")```
-->
```print("   " + " OK - CRPS CDF plot saved")```

File: coast/NEMO.py [hunk 5]
```diff
@@ -278,6 +279,129 @@ def transect_indices(self, start: tuple, end: tuple) -> tuple:
...
+    
+    def interpolate_in_time(self, model_array, new_times, 
+                               interp_method = 'nearest', extrapolate=True):
+        '''
+        Interpolates a provided xarray.DataArray in time to new python
+        datetimes using a specified scipy.interpolate method.
+        
+        Example Useage
+        ----------
+
```
Comment (jpolton) [inline coast/NEMO.py:377]: Add something to Example Usage. This sort of thing is really worth the effort. Then add to unit test

File: coast/NEMO.py [hunk 6]
```diff
@@ -278,6 +279,129 @@ def transect_indices(self, start: tuple, end: tuple) -> tuple:
...
+    def nearest_xy_indices(self, model_dataset, new_lons, new_lats):
+        '''
+        Obtains the x and y indices of the nearest model points to specified
+        lists of longitudes and latitudes. Makes use of sklearn.neighbours
+        and its BallTree haversine method. 
+        
+        Example Useage
+        ----------
+        # Get indices of model points closest to altimetry points
+        ind_x, ind_y = nemo.nearest_indices(altimetry.dataset.longitude,
```
Comment (jpolton) [inline coast/NEMO.py:291]: Add this to the unit test

File: coast/NEMO.py [hunk 7]
```diff
@@ -278,6 +279,129 @@ def transect_indices(self, start: tuple, end: tuple) -> tuple:
...
+    
+    def interpolate_in_space(self, model_array, new_lons, new_lats):
+        '''
+        Interpolates a provided xarray.DataArray in space to new longitudes
+        and latitudes using a nearest neighbour method (BallTree).
+        
+        Example Useage
+        ----------
+        # Get an interpolated DataArray for temperature onto two locations
+        interpolated = nemo.interpolate_in_space(nemo.dataset.votemper,
```
Comment (jpolton) [inline coast/NEMO.py:348]: Add to unit test

File: unit_testing/unit_test.py (hunk 4 above)
Comment (davbyr) [reply unit_testing/unit_test.py:562]: Separated into two units

File: coast/NEMO.py (hunk 5 above)
Comment (davbyr) [reply coast/NEMO.py:377]: @jpolton Done. So is the idea to put example useage in every method?

File: coast/NEMO.py (hunk 6 above)
Comment (davbyr) [reply coast/NEMO.py:291]: Done

File: unit_testing/unit_test.py (hunk 4 above)
Comment (anwiseNOCL) [reply unit_testing/unit_test.py:562]: just need to update the 6b to 6c in the text.

File: coast/NEMO.py (hunk 5 above)
Comment (jpolton) [reply coast/NEMO.py:377]: Maybe. I find from experience that if a method has an example. I am more likely to try it. It provides an alternative route to understand the method and a rapid route to adoption.

**Comment by davbyr** [comment]
@jpolton @anwiseNOCL I ahve updated this with the changes you suggested. The unit testing has changed a fair bit so it might be worth rerunning it to make sure it still works for you two.

**Pull request merged**